Sources/Hummingbird/Router/BinaryTrie/BinaryTrie+resolve.swift

```diff
@@ -0,0 +1,246 @@
+        if components.isEmpty {
+            return value(for: index, parameters: parameters)
+        }
+
+        // Take the next component from the path
+        var component = components.removeFirst()
+        
+        // Check the current node type through TokenKind
+        // And read the location of the _next_ node from the trie buffer
+        while 
+            let index = trie.readInteger(as: UInt16.self),
+            let _token: Integer = trie.readInteger(),
+            let token = TokenKind(rawValue: _token),
+            let nextSiblingNodeIndex: UInt32 = trie.readInteger()
+        {
+            repeat {
+                // Record the current readerIndex
+                // ``matchComponent`` moves the reader index forward, so we'll need to reset it
+                // If we're in a recursiveWildcard and this component does not match
+                let readerIndex = trie.readerIndex
+                let result = matchComponent(&component, withToken: token, in: &trie, parameters: &parameters)
+
+                switch result {
+                case .match:
+                    return descendPath(
+                        in: &trie,
+                        index: index,
+                        parameters: &parameters,
+                        components: &components,
+                        isInRecursiveWildcard: false
+                    )
+                case .mismatch where isInRecursiveWildcard:
+                    if components.isEmpty {
+                        return nil
+                    }
+
+                    component = components.removeFirst()
+                    // Move back he readerIndex, so that we can retry this step again with
+                    // the next component
+                    trie.moveReaderIndex(to: readerIndex)
+                case .mismatch:
+                    // Move to the next sibling-node, not descending a level
+                    trie.moveReaderIndex(to: Int(nextSiblingNodeIndex))
+                    continue
+                case .recursivelyDiscarded:
+                    return descendPath(
+                        in: &trie,
+                        index: index,
+                        parameters: &parameters,
+                        components: &components,
+                        isInRecursiveWildcard: true
+                    )
+                case .ignore:
+                    continue
+                case .deadEnd:
+                    return nil
+                }
+            } while isInRecursiveWildcard
+        }
+
+        return nil
+    }
+}
+
+#if canImport(Darwin)
+import Darwin.C
+#elseif canImport(Musl)
+import Musl
+#elseif os(Linux) || os(FreeBSD) || os(Android)
+import Glibc
+#else
+#error("unsupported os")
+#endif
+
+fileprivate extension ByteBuffer {
+    mutating func readAndCompareString<Length: FixedWidthInteger>(to string: inout Substring, length: Length) -> Bool {
```

Comment:
This shouldn't require an `inout Substring`. Use `withContiguousStorageIfAvailable` instead and if it fails (which is rare) copy string and call `makeContiguousUTF8` on copy and then call function with the string copy